# 11 Visualization Preparation

## Purpose

This notebook prepares the final analytical datasets for visualization and presentation.

No new statistical tests, predictive models, or strategic classifications are introduced. Instead, the notebook consolidates the validated outputs from the previous stages into a small set of visualization-ready tables.

The final visual analysis is designed to answer the project's central business question:

**How competitive is Germany's tourism sector compared with other EU destinations, what factors explain its current position, and where are the strongest opportunities for future tourism growth?**

## Visualization Principles

The final dashboards and presentation should:

- distinguish tourism scale from tourism depth
- emphasize insight rather than display every available metric
- maintain consistent definitions across charts
- clearly distinguish actual data from projections
- identify source-market opportunities without implying causality
- treat counterfactual scenarios as illustrative benchmarks
- preserve data-quality and comparability limitations
- connect every visualization to a business implication

The final Tableau story will therefore follow the analytical sequence:

**Scale → Tourism Depth → Source Markets → Current Momentum → Outlook → Strategy**

In [18]:
import pandas as pd
import numpy as np

from pathlib import Path

pd.set_option(
    "display.max_columns",
    None
)

pd.set_option(
    "display.max_rows",
    100
)

project_root = Path.cwd().parent

processed_dir = (
    project_root
    / "data"
    / "processed"
)

tableau_dir = (
    project_root
    / "tableau"
)

tableau_dir.mkdir(
    parents=True,
    exist_ok=True
)

print(
    "Tableau directory:",
    tableau_dir
)

Tableau directory: /Users/jannoelvero/Documents/germany_eu_tourism_analysis/tableau


In [19]:
eu_benchmark = pd.read_csv(
    processed_dir
    / "eu27_tourism_benchmark_2024.csv"
)

eu_annual = pd.read_csv(
    processed_dir
    / "eu27_tourism_benchmark_annual.csv"
)

los_scenarios = pd.read_csv(
    processed_dir
    / "germany_los_counterfactual_scenarios_2024.csv"
)

source_markets = pd.read_csv(
    processed_dir
    / "germany_source_market_action_matrix.csv"
)

ytd_monthly = pd.read_csv(
    processed_dir
    / "germany_2026_ytd_monthly_comparison.csv"
)

actual_projection = pd.read_csv(
    processed_dir
    / "germany_2026_actual_projection_combined.csv"
)

economic_context = pd.read_csv(
    processed_dir
    / "germany_strategic_economic_context.csv"
)

recommendations = pd.read_csv(
    processed_dir
    / "germany_strategic_recommendations.csv"
)

print(
    "EU 2024 benchmark:",
    eu_benchmark.shape
)

print(
    "EU annual benchmark:",
    eu_annual.shape
)

print(
    "LOS scenarios:",
    los_scenarios.shape
)

print(
    "Source markets:",
    source_markets.shape
)

print(
    "2026 monthly comparison:",
    ytd_monthly.shape
)

print(
    "2026 actual/projection:",
    actual_projection.shape
)

print(
    "Economic context:",
    economic_context.shape
)

print(
    "Recommendations:",
    recommendations.shape
)

EU 2024 benchmark: (27, 11)
EU annual benchmark: (123, 8)
LOS scenarios: (6, 5)
Source markets: (16, 10)
2026 monthly comparison: (6, 13)
2026 actual/projection: (12, 5)
Economic context: (5, 8)
Recommendations: (4, 5)


In [20]:
visualization_inputs = {
    "EU 2024 Benchmark":
        eu_benchmark,

    "EU Annual Benchmark":
        eu_annual,

    "LOS Scenarios":
        los_scenarios,

    "Source Markets":
        source_markets,

    "2026 Monthly Comparison":
        ytd_monthly,

    "2026 Actual Projection":
        actual_projection,

    "Economic Context":
        economic_context,

    "Recommendations":
        recommendations
}

for name, df in visualization_inputs.items():

    print(
        f"\n{name}"
    )

    print(
        df.columns.tolist()
    )

    display(
        df.head()
    )


EU 2024 Benchmark
['country_code', 'country', 'foreign_arrivals', 'arrivals_rank', 'foreign_nights', 'nights_rank', 'avg_length_of_stay', 'los_rank', 'scale_position', 'stay_position', 'competitive_position']


,country_code,country,foreign_arrivals,arrivals_rank,foreign_nights,nights_rank,avg_length_of_stay,los_rank,scale_position,stay_position,competitive_position
0,AT,Austria,28498633.0,5,94009113.0,5,3.298724,10,High Scale,Longer Stay,High Scale / Longer Stay
1,BE,Belgium,9638187.0,11,22279289.0,12,2.311564,20,High Scale,Shorter Stay,High Scale / Shorter Stay
2,BG,Bulgaria,2994009.0,19,15058045.0,18,5.029392,3,Low Scale,Longer Stay,Low Scale / Longer Stay
3,CY,Cyprus,3030161.0,18,16558854.0,15,5.464678,2,Low Scale,Longer Stay,Low Scale / Longer Stay
4,CZ,Czechia,10477528.0,10,25474111.0,10,2.431309,16,High Scale,Shorter Stay,High Scale / Shorter Stay



EU Annual Benchmark
['country_code', 'country', 'year', 'foreign_arrivals', 'foreign_nights', 'avg_length_of_stay', 'los_rank', 'countries_compared']


,country_code,country,year,foreign_arrivals,foreign_nights,avg_length_of_stay,los_rank,countries_compared
0,AT,Austria,2021,11098084.0,40034274.0,3.607314,8,22
1,AT,Austria,2022,22874333.0,79470210.0,3.474209,8,24
2,AT,Austria,2023,27212813.0,91023501.0,3.344877,9,24
3,AT,Austria,2024,28498633.0,94009113.0,3.298724,10,27
4,AT,Austria,2025,29764173.0,97020394.0,3.259637,9,26



LOS Scenarios
['benchmark', 'benchmark_los', 'illustrative_nights', 'additional_nights_vs_actual', 'additional_nights_pct']


,benchmark,benchmark_los,illustrative_nights,additional_nights_vs_actual,additional_nights_pct
0,Germany Actual,2.265983,8.479264e+07,0.000000e+00,0.000000
1,EU27 Median,2.490083,9.317841e+07,8.385770e+06,9.889738
2,France,2.577033,9.643208e+07,1.163944e+07,13.726946
3,Netherlands,2.894230,1.083015e+08,2.350887e+07,27.725134
4,Austria,3.298724,1.234376e+08,3.864494e+07,45.575818



Source Markets
['source_market_code', 'source_market', 'foreign_nights', 'germany_foreign_nights_share_pct', 'growth_2023_2024_pct', 'absolute_growth_nights', 'avg_length_of_stay', 'strategic_role', 'stay_extension_priority', 'recommended_action']


,source_market_code,source_market,foreign_nights,germany_foreign_nights_share_pct,growth_2023_2024_pct,absolute_growth_nights,avg_length_of_stay,strategic_role,stay_extension_priority,recommended_action
0,US,United States,7223819.0,8.519394,10.750402,701207.0,2.241385,Core Growth Market,Lower,Protect growth while strengthening repeat visi...
1,UK,United Kingdom,5274041.0,6.219928,10.880105,517515.0,2.222397,Core Growth Market,Lower,Protect growth while strengthening repeat visi...
2,NL,Netherlands,11867612.0,13.996041,3.659706,418986.0,2.260446,Core Retention Market,Lower,Defend market share and maintain strong visito...
3,AT,Austria,4360393.0,5.142419,4.927288,204760.0,2.045902,Stay Extension Opportunity,High,"Prioritize longer-stay products, multi-destina..."
4,PL,Poland,4167089.0,4.914447,0.155626,6475.0,2.832105,Core Retention Market,Lower,Defend market share and maintain strong visito...



2026 Monthly Comparison
['month', 'arrivals_2025', 'arrivals_2026', 'arrivals_change', 'arrivals_growth_pct', 'nights_2025', 'nights_2026', 'nights_change', 'nights_growth_pct', 'los_2025', 'los_2026', 'los_change', 'los_change_pct']


,month,arrivals_2025,arrivals_2026,arrivals_change,arrivals_growth_pct,nights_2025,nights_2026,nights_change,nights_growth_pct,los_2025,los_2026,los_change,los_change_pct
0,1,1869061.0,1938956.0,69895.0,3.739578,4242218.0,4277203.0,34985.0,0.824687,2.269705,2.205931,-0.063775,-2.809817
1,2,2180842.0,2271693.0,90851.0,4.165868,4949805.0,5078278.0,128473.0,2.595516,2.269676,2.235460,-0.034216,-1.507549
2,3,2294956.0,2341823.0,46867.0,2.042174,5220357.0,5197078.0,-23279.0,-0.445927,2.274709,2.219245,-0.055464,-2.438307
3,4,2898714.0,2934984.0,36270.0,1.251245,6556296.0,6596278.0,39982.0,0.609826,2.261795,2.247466,-0.014328,-0.633492
4,5,3329016.0,3353419.0,24403.0,0.733039,7484744.0,7446648.0,-38096.0,-0.508982,2.248335,2.220614,-0.027722,-1.232983



2026 Actual Projection
['date', 'year', 'month', 'foreign_nights_value', 'data_type']


,date,year,month,foreign_nights_value,data_type
0,2026-01-01,2026,1,4277203.0,Actual
1,2026-02-01,2026,2,5078278.0,Actual
2,2026-03-01,2026,3,5197078.0,Actual
3,2026-04-01,2026,4,6596278.0,Actual
4,2026-05-01,2026,5,7446648.0,Actual



Economic Context
['year', 'travel_receipts_mio_eur', 'travel_expenditure_mio_eur', 'travel_balance_mio_eur', 'receipts_cover_expenditure_pct', 'receipts_growth_pct', 'expenditure_growth_pct', 'is_provisional']


,year,travel_receipts_mio_eur,travel_expenditure_mio_eur,travel_balance_mio_eur,receipts_cover_expenditure_pct,receipts_growth_pct,expenditure_growth_pct,is_provisional
0,2021,18827.0,43126.0,-24300.0,43.655799,NaN,NaN,False
1,2022,30257.0,85019.0,-54762.0,35.588516,60.710681,97.140936,False
2,2023,34992.0,106642.0,-71650.0,32.812588,15.649271,25.433138,False
3,2024,37055.0,106822.0,-69767.0,34.688547,5.895633,0.168789,False
4,2025,37772.0,114594.0,-76822.0,32.961586,1.934962,7.275655,True



Recommendations
['priority', 'strategic_priority', 'evidence_basis', 'recommended_direction', 'primary_business_kpi']


,priority,strategic_priority,evidence_basis,recommended_direction,primary_business_kpi
0,1,Increase Stay Duration,Germany ranks 21st in EU27 LOS despite ranking...,"Develop trip-extension, multi-city and regiona...",Average Length of Stay
1,2,Differentiate Source-Market Strategy,Major source markets differ substantially in s...,"Separate growth, retention, stay-extension and...",Foreign Nights by Source Market
2,3,Convert Growth into Tourism Depth,H1 2026 arrivals grew faster than nights while...,Evaluate success using nights and LOS alongsid...,Nights Growth Relative to Arrivals Growth
3,4,Monitor Performance with Leading Indicators,Seasonal persistence is strong and the 2026 ba...,"Track monthly arrivals, nights, LOS and source...",Monthly Forecast Error and YTD Performance


In [21]:
tableau_eu_competitive_position = (
    eu_benchmark
    .copy()
)

tableau_eu_competitive_position[
    "germany_flag"
] = np.where(
    tableau_eu_competitive_position[
        "country_code"
    ] == "DE",
    "Germany",
    "Other EU27"
)

tableau_eu_competitive_position[
    "year"
] = 2024

tableau_eu_competitive_position = (
    tableau_eu_competitive_position[
        [
            "year",
            "country_code",
            "country",
            "foreign_arrivals",
            "arrivals_rank",
            "foreign_nights",
            "nights_rank",
            "avg_length_of_stay",
            "los_rank",
            "scale_position",
            "stay_position",
            "competitive_position",
            "germany_flag"
        ]
    ]
)

display(
    tableau_eu_competitive_position.head()
)

display(
    tableau_eu_competitive_position[
        tableau_eu_competitive_position[
            "country_code"
        ] == "DE"
    ]
)

,year,country_code,country,foreign_arrivals,arrivals_rank,foreign_nights,nights_rank,avg_length_of_stay,los_rank,scale_position,stay_position,competitive_position,germany_flag
0,2024,AT,Austria,28498633.0,5,94009113.0,5,3.298724,10,High Scale,Longer Stay,High Scale / Longer Stay,Other EU27
1,2024,BE,Belgium,9638187.0,11,22279289.0,12,2.311564,20,High Scale,Shorter Stay,High Scale / Shorter Stay,Other EU27
2,2024,BG,Bulgaria,2994009.0,19,15058045.0,18,5.029392,3,Low Scale,Longer Stay,Low Scale / Longer Stay,Other EU27
3,2024,CY,Cyprus,3030161.0,18,16558854.0,15,5.464678,2,Low Scale,Longer Stay,Low Scale / Longer Stay,Other EU27
4,2024,CZ,Czechia,10477528.0,10,25474111.0,10,2.431309,16,High Scale,Shorter Stay,High Scale / Shorter Stay,Other EU27


,year,country_code,country,foreign_arrivals,arrivals_rank,foreign_nights,nights_rank,avg_length_of_stay,los_rank,scale_position,stay_position,competitive_position,germany_flag
5,2024,DE,Germany,37419799.0,4,84792636.0,7,2.265983,21,High Scale,Shorter Stay,High Scale / Shorter Stay,Germany


In [22]:
tableau_eu_los = (
    eu_benchmark[
        [
            "country_code",
            "country",
            "avg_length_of_stay",
            "los_rank"
        ]
    ]
    .copy()
)

tableau_eu_los[
    "year"
] = 2024

tableau_eu_los[
    "germany_flag"
] = np.where(
    tableau_eu_los[
        "country_code"
    ] == "DE",
    "Germany",
    "Other EU27"
)

tableau_eu_los = (
    tableau_eu_los
    .sort_values(
        "avg_length_of_stay",
        ascending=False
    )
    .reset_index(
        drop=True
    )
)

display(
    tableau_eu_los
)

,country_code,country,avg_length_of_stay,los_rank,year,germany_flag
0,IE,Ireland,5.792956,1,2024,Other EU27
1,CY,Cyprus,5.464678,2,2024,Other EU27
2,BG,Bulgaria,5.029392,3,2024,Other EU27
3,HR,Croatia,4.888926,4,2024,Other EU27
4,EL,Greece,4.682106,5,2024,Other EU27
5,MT,Malta,4.654179,6,2024,Other EU27
6,DK,Denmark,4.373966,7,2024,Other EU27
7,ES,Spain,4.149108,8,2024,Other EU27
8,IT,Italy,3.433996,9,2024,Other EU27
9,AT,Austria,3.298724,10,2024,Other EU27


In [23]:
tableau_los_scenarios = (
    los_scenarios
    .copy()
)

tableau_los_scenarios[
    "scenario_type"
] = np.where(
    tableau_los_scenarios[
        "benchmark"
    ] == "Germany Actual",
    "Actual",
    "Illustrative Counterfactual"
)

display(
    tableau_los_scenarios
)

,benchmark,benchmark_los,illustrative_nights,additional_nights_vs_actual,additional_nights_pct,scenario_type
0,Germany Actual,2.265983,8.479264e+07,0.000000e+00,0.000000,Actual
1,EU27 Median,2.490083,9.317841e+07,8.385770e+06,9.889738,Illustrative Counterfactual
2,France,2.577033,9.643208e+07,1.163944e+07,13.726946,Illustrative Counterfactual
3,Netherlands,2.894230,1.083015e+08,2.350887e+07,27.725134,Illustrative Counterfactual
4,Austria,3.298724,1.234376e+08,3.864494e+07,45.575818,Illustrative Counterfactual
5,Weighted EU27,3.340360,1.249956e+08,4.020298e+07,47.413287,Illustrative Counterfactual


In [24]:
tableau_source_markets = (
    source_markets
    .copy()
)

tableau_source_markets[
    "market_label"
] = (
    tableau_source_markets[
        "source_market"
    ]
    + " ("
    + tableau_source_markets[
        "source_market_code"
    ]
    + ")"
)

priority_order = {
    "High": 1,
    "Emerging": 2,
    "Lower": 3
}

tableau_source_markets[
    "stay_extension_priority_order"
] = (
    tableau_source_markets[
        "stay_extension_priority"
    ]
    .map(
        priority_order
    )
)

display(
    tableau_source_markets[
        [
            "source_market_code",
            "source_market",
            "foreign_nights",
            "growth_2023_2024_pct",
            "avg_length_of_stay",
            "strategic_role",
            "stay_extension_priority",
            "stay_extension_priority_order"
        ]
    ]
)

,source_market_code,source_market,foreign_nights,growth_2023_2024_pct,avg_length_of_stay,strategic_role,stay_extension_priority,stay_extension_priority_order
0,US,United States,7223819.0,10.750402,2.241385,Core Growth Market,Lower,3
1,UK,United Kingdom,5274041.0,10.880105,2.222397,Core Growth Market,Lower,3
2,NL,Netherlands,11867612.0,3.659706,2.260446,Core Retention Market,Lower,3
3,AT,Austria,4360393.0,4.927288,2.045902,Stay Extension Opportunity,High,1
4,PL,Poland,4167089.0,0.155626,2.832105,Core Retention Market,Lower,3
5,FR,France,3680328.0,4.836640,1.990412,Stay Extension Opportunity,High,1
6,IT,Italy,3497676.0,4.605835,2.217334,Stay Extension Opportunity,High,1
7,ES,Spain,2720461.0,7.653031,2.422259,Emerging Growth Market,Lower,3
8,TR,Türkiye,1225934.0,17.945975,2.510694,Emerging Growth Market,Lower,3
9,CH_LI,Switzerland and Liechtenstein,6897129.0,2.414773,2.080067,Monitor / Maintain,High,1


In [25]:
tableau_ytd_growth = (
    ytd_monthly[
        [
            "month",
            "arrivals_growth_pct",
            "nights_growth_pct",
            "los_change_pct"
        ]
    ]
    .melt(
        id_vars=[
            "month"
        ],
        value_vars=[
            "arrivals_growth_pct",
            "nights_growth_pct",
            "los_change_pct"
        ],
        var_name="indicator",
        value_name="change_pct"
    )
)

indicator_labels = {
    "arrivals_growth_pct":
        "Foreign Arrivals",

    "nights_growth_pct":
        "Foreign Nights",

    "los_change_pct":
        "Average Length of Stay"
}

tableau_ytd_growth[
    "indicator"
] = (
    tableau_ytd_growth[
        "indicator"
    ]
    .map(
        indicator_labels
    )
)

tableau_ytd_growth[
    "month_name"
] = pd.to_datetime(
    tableau_ytd_growth[
        "month"
    ],
    format="%m"
).dt.month_name()

tableau_ytd_growth[
    "period"
] = "2026 vs 2025"

display(
    tableau_ytd_growth
)

,month,indicator,change_pct,month_name,period
0,1,Foreign Arrivals,3.739578,January,2026 vs 2025
1,2,Foreign Arrivals,4.165868,February,2026 vs 2025
2,3,Foreign Arrivals,2.042174,March,2026 vs 2025
3,4,Foreign Arrivals,1.251245,April,2026 vs 2025
4,5,Foreign Arrivals,0.733039,May,2026 vs 2025
5,6,Foreign Arrivals,1.676406,June,2026 vs 2025
6,1,Foreign Nights,0.824687,January,2026 vs 2025
7,2,Foreign Nights,2.595516,February,2026 vs 2025
8,3,Foreign Nights,-0.445927,March,2026 vs 2025
9,4,Foreign Nights,0.609826,April,2026 vs 2025


In [26]:
tableau_2026_outlook = (
    actual_projection
    .copy()
)

tableau_2026_outlook[
    "date"
] = pd.to_datetime(
    tableau_2026_outlook[
        "date"
    ]
)

tableau_2026_outlook[
    "observation_status"
] = np.where(
    tableau_2026_outlook[
        "data_type"
    ] == "Actual",
    "Observed",
    "Projected"
)

display(
    tableau_2026_outlook
)

,date,year,month,foreign_nights_value,data_type,observation_status
0,2026-01-01,2026,1,4277203.0,Actual,Observed
1,2026-02-01,2026,2,5078278.0,Actual,Observed
2,2026-03-01,2026,3,5197078.0,Actual,Observed
3,2026-04-01,2026,4,6596278.0,Actual,Observed
4,2026-05-01,2026,5,7446648.0,Actual,Observed
5,2026-06-01,2026,6,7653746.0,Actual,Observed
6,2026-07-01,2026,7,10358796.0,Seasonal Naive Projection,Projected
7,2026-08-01,2026,8,9689151.0,Seasonal Naive Projection,Projected
8,2026-09-01,2026,9,7674317.0,Seasonal Naive Projection,Projected
9,2026-10-01,2026,10,7195093.0,Seasonal Naive Projection,Projected


In [27]:
tableau_economic_context = (
    economic_context
    .copy()
)

tableau_economic_context[
    "data_status"
] = np.where(
    tableau_economic_context[
        "is_provisional"
    ],
    "Provisional",
    "Final"
)

display(
    tableau_economic_context
)

,year,travel_receipts_mio_eur,travel_expenditure_mio_eur,travel_balance_mio_eur,receipts_cover_expenditure_pct,receipts_growth_pct,expenditure_growth_pct,is_provisional,data_status
0,2021,18827.0,43126.0,-24300.0,43.655799,NaN,NaN,False,Final
1,2022,30257.0,85019.0,-54762.0,35.588516,60.710681,97.140936,False,Final
2,2023,34992.0,106642.0,-71650.0,32.812588,15.649271,25.433138,False,Final
3,2024,37055.0,106822.0,-69767.0,34.688547,5.895633,0.168789,False,Final
4,2025,37772.0,114594.0,-76822.0,32.961586,1.934962,7.275655,True,Provisional


In [28]:
tableau_recommendations = (
    recommendations
    .sort_values(
        "priority"
    )
    .reset_index(
        drop=True
    )
    .copy()
)

display(
    tableau_recommendations
)

,priority,strategic_priority,evidence_basis,recommended_direction,primary_business_kpi
0,1,Increase Stay Duration,Germany ranks 21st in EU27 LOS despite ranking...,"Develop trip-extension, multi-city and regiona...",Average Length of Stay
1,2,Differentiate Source-Market Strategy,Major source markets differ substantially in s...,"Separate growth, retention, stay-extension and...",Foreign Nights by Source Market
2,3,Convert Growth into Tourism Depth,H1 2026 arrivals grew faster than nights while...,Evaluate success using nights and LOS alongsid...,Nights Growth Relative to Arrivals Growth
3,4,Monitor Performance with Leading Indicators,Seasonal persistence is strong and the 2026 ba...,"Track monthly arrivals, nights, LOS and source...",Monthly Forecast Error and YTD Performance


In [29]:
tableau_recommendations = (
    recommendations
    .sort_values(
        "priority"
    )
    .reset_index(
        drop=True
    )
    .copy()
)

display(
    tableau_recommendations
)

,priority,strategic_priority,evidence_basis,recommended_direction,primary_business_kpi
0,1,Increase Stay Duration,Germany ranks 21st in EU27 LOS despite ranking...,"Develop trip-extension, multi-city and regiona...",Average Length of Stay
1,2,Differentiate Source-Market Strategy,Major source markets differ substantially in s...,"Separate growth, retention, stay-extension and...",Foreign Nights by Source Market
2,3,Convert Growth into Tourism Depth,H1 2026 arrivals grew faster than nights while...,Evaluate success using nights and LOS alongsid...,Nights Growth Relative to Arrivals Growth
3,4,Monitor Performance with Leading Indicators,Seasonal persistence is strong and the 2026 ba...,"Track monthly arrivals, nights, LOS and source...",Monthly Forecast Error and YTD Performance


In [30]:
print(
    "EU competitive position rows:",
    len(tableau_eu_competitive_position)
)

print(
    "Germany rows:",
    (
        tableau_eu_competitive_position[
            "germany_flag"
        ] == "Germany"
    ).sum()
)

print(
    "EU LOS rows:",
    len(tableau_eu_los)
)

print(
    "LOS scenario rows:",
    len(tableau_los_scenarios)
)

print(
    "Counterfactual scenarios:",
    (
        tableau_los_scenarios[
            "scenario_type"
        ] == "Illustrative Counterfactual"
    ).sum()
)

print(
    "Source market rows:",
    len(tableau_source_markets)
)

print(
    "YTD growth rows:",
    len(tableau_ytd_growth)
)

print(
    "2026 outlook rows:",
    len(tableau_2026_outlook)
)

print(
    "Observed months:",
    (
        tableau_2026_outlook[
            "observation_status"
        ] == "Observed"
    ).sum()
)

print(
    "Projected months:",
    (
        tableau_2026_outlook[
            "observation_status"
        ] == "Projected"
    ).sum()
)

print(
    "Provisional economic rows:",
    (
        tableau_economic_context[
            "data_status"
        ] == "Provisional"
    ).sum()
)

print(
    "Recommendation rows:",
    len(tableau_recommendations)
)

EU competitive position rows: 27
Germany rows: 1
EU LOS rows: 27
LOS scenario rows: 6
Counterfactual scenarios: 5
Source market rows: 16
YTD growth rows: 18
2026 outlook rows: 12
Observed months: 6
Projected months: 6
Provisional economic rows: 1
Recommendation rows: 4


In [31]:
tableau_exports = {
    "tableau_eu_competitive_position.csv":
        tableau_eu_competitive_position,

    "tableau_eu_los_2024.csv":
        tableau_eu_los,

    "tableau_los_scenarios_2024.csv":
        tableau_los_scenarios,

    "tableau_source_market_strategy_2024.csv":
        tableau_source_markets,

    "tableau_2026_ytd_growth.csv":
        tableau_ytd_growth,

    "tableau_2026_outlook.csv":
        tableau_2026_outlook,

    "tableau_economic_context.csv":
        tableau_economic_context,

    "tableau_strategic_recommendations.csv":
        tableau_recommendations
}

for file_name, df in tableau_exports.items():

    output_path = (
        tableau_dir
        / file_name
    )

    df.to_csv(
        output_path,
        index=False
    )

    print(
        file_name,
        "->",
        df.shape
    )

tableau_eu_competitive_position.csv -> (27, 13)
tableau_eu_los_2024.csv -> (27, 6)
tableau_los_scenarios_2024.csv -> (6, 6)
tableau_source_market_strategy_2024.csv -> (16, 12)
tableau_2026_ytd_growth.csv -> (18, 5)
tableau_2026_outlook.csv -> (12, 6)
tableau_economic_context.csv -> (5, 9)
tableau_strategic_recommendations.csv -> (4, 5)


In [32]:
for file_name in tableau_exports:

    file_path = (
        tableau_dir
        / file_name
    )

    print(
        file_name,
        "->",
        file_path.exists()
    )

tableau_eu_competitive_position.csv -> True
tableau_eu_los_2024.csv -> True
tableau_los_scenarios_2024.csv -> True
tableau_source_market_strategy_2024.csv -> True
tableau_2026_ytd_growth.csv -> True
tableau_2026_outlook.csv -> True
tableau_economic_context.csv -> True
tableau_strategic_recommendations.csv -> True


# Tableau Visualization Architecture

The final visualization is designed as a business story rather than a collection of independent charts.

## Story Point 1: Germany Is a Major EU Tourism Destination

**Business question:** How strong is Germany's international tourism position?

**Primary dataset:** `tableau_eu_competitive_position.csv`

**Primary visual:** EU27 competitive-position scatter plot

- X-axis: Foreign Arrivals
- Y-axis: Foreign Nights
- Detail: Country
- Highlight: Germany
- Tooltip: Arrivals Rank, Nights Rank, Average Length of Stay

**Key message:** Germany has substantial international tourism scale, ranking fourth in arrivals and seventh in overnight stays in the complete 2024 EU27 comparison.

---

## Story Point 2: The Competitive Gap Is Tourism Depth

**Business question:** Why does Germany's strong visitor scale not translate into a higher overnight-stay position?

**Primary datasets:**
- `tableau_eu_los_2024.csv`
- `tableau_los_scenarios_2024.csv`

**Primary visuals:**
- EU27 average-length-of-stay ranking
- Illustrative LOS benchmark scenarios

**Key message:** Germany's average international stay is approximately 2.27 nights, ranking 21st among the EU27 in 2024.

The EU27 median scenario illustrates that even a moderate increase in stay duration could correspond to materially greater overnight-stay volume at Germany's existing arrival scale.

**Important:** Counterfactual scenarios are illustrative benchmarks, not forecasts or estimates of recoverable demand.

---

## Story Point 3: Different Markets Require Different Strategies

**Business question:** Where are Germany's strongest source-market opportunities?

**Primary dataset:** `tableau_source_market_strategy_2024.csv`

**Primary visual:** Source-market opportunity scatter plot

- X-axis: 2023–2024 Growth %
- Y-axis: Average Length of Stay
- Size: Foreign Nights
- Detail: Source Market
- Strategic segmentation: Strategic Role

**Key message:** Germany should not use a uniform international acquisition strategy. Established growth, retention, stay-extension, emerging-growth and long-stay markets require different objectives.

---

## Story Point 4: 2026 Growth Is Not Translating Fully into Tourism Depth

**Business question:** Is Germany's current tourism performance strengthening?

**Primary dataset:** `tableau_2026_ytd_growth.csv`

**Primary visual:** Monthly year-over-year percentage-change comparison

- Month: January–June
- Measure: Change %
- Series: Foreign Arrivals, Foreign Nights, Average Length of Stay

**Key message:** International arrivals increased throughout the first half of 2026, but overnight stays grew more slowly and average length of stay declined.

---

## Story Point 5: Near-Term Overnight-Stay Growth Remains Modest

**Business question:** What does the validated baseline indicate for the remainder of 2026?

**Primary datasets:**
- `tableau_2026_outlook.csv`
- `tableau_economic_context.csv`

**Primary visuals:**
- 2026 observed versus seasonal-naive baseline line chart
- Travel receipts and expenditure context

**Key message:** The seasonal-naive baseline suggests limited full-year overnight-stay growth. Inbound travel receipts have recovered strongly, but economic and tourism indicators should be interpreted within their separate statistical scopes.

**Important:** January–June values are observed. July–December values are seasonal-naive projections.

---

## Story Point 6: Shift from Volume-Only Growth to Tourism Depth

**Business question:** What should tourism decision-makers prioritize?

**Primary dataset:** `tableau_strategic_recommendations.csv`

**Primary visual:** Strategic recommendation and KPI summary

**Key message:** Germany should protect its strong international visitor base while increasing tourism depth through stay extension, differentiated source-market strategies, nights-based performance evaluation and continuous seasonal monitoring.

## Final Story

**Germany's tourism challenge is not primarily attracting international visitors. It is converting an already strong visitor position into greater overnight-stay depth and tourism value.**

# Conclusion

The visualization-preparation stage consolidates the project's validated analytical outputs into a focused set of Tableau-ready datasets.

The final visualization architecture follows the analytical logic established throughout the project:

**Scale → Tourism Depth → Source Markets → Current Momentum → Outlook → Strategy**

This structure ensures that the final dashboards do more than report tourism indicators. Each visualization contributes directly to answering the central business question and leads logically toward the project's strategic recommendations.

No new analytical findings are introduced at this stage. The purpose of the visualization layer is to communicate the existing evidence clearly, accurately and consistently while preserving important distinctions between observed data, statistical findings, counterfactual benchmarks, predictive scenarios and strategic recommendations.

In [36]:
clean_dir = (
    project_root
    / "data"
    / "clean"
)

print(
    "Clean data directory:",
    clean_dir
)

Clean data directory: /Users/jannoelvero/Documents/germany_eu_tourism_analysis/data/clean


In [37]:
country_reference = pd.read_csv(
    clean_dir
    / "eu27_country_reference_clean.csv"
)

print(country_reference.shape)
print(country_reference.columns.tolist())

display(
    country_reference.head()
)

(27, 2)
['country_code', 'country']


,country_code,country
0,AT,Austria
1,BE,Belgium
2,BG,Bulgaria
3,HR,Croatia
4,CY,Cyprus


In [38]:
source_market_reference = pd.read_csv(
    clean_dir
    / "source_market_reference_clean.csv"
)

print(source_market_reference.shape)
print(source_market_reference.columns.tolist())

display(
    source_market_reference.head(10)
)

(63, 4)
['source_market_code', 'source_market', 'market_type', 'international_market']


,source_market_code,source_market,market_type,international_market
0,AFR,Africa,regional_aggregate,False
1,AFR_OTH,Other African countries (aggregate changing ac...,regional_aggregate,False
2,AME,America,regional_aggregate,False
3,AME_C_S,Central and South America,regional_aggregate,False
4,AME_C_S_OTH,Other Central or South American countries,regional_aggregate,False
5,AME_N,Northern America,regional_aggregate,False
6,AME_N_OTH,Other Northern American countries,regional_aggregate,False
7,ASI,Asia,regional_aggregate,False
8,ASI_OTH,Other Asian countries (aggregate changing acco...,regional_aggregate,False
9,AT,Austria,individual_country,True


In [39]:
from sqlalchemy import create_engine

In [42]:
engine = create_engine(
    "mysql+pymysql://root:JanMysql2026!@127.0.0.1:3306/germany_eu_tourism"
)

In [43]:
from sqlalchemy import text

with engine.connect() as connection:
    result = connection.execute(
        text("SELECT DATABASE();")
    )

    print(result.fetchone())

('germany_eu_tourism',)


In [44]:
country_sql = country_reference.copy()

country_sql = country_sql.rename(
    columns={
        "country": "country_name"
    }
)

country_sql["is_germany"] = (
    country_sql["country_code"] == "DE"
)

country_sql.head()

,country_code,country_name,is_germany
0,AT,Austria,False
1,BE,Belgium,False
2,BG,Bulgaria,False
3,HR,Croatia,False
4,CY,Cyprus,False


In [45]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                COUNT(*) AS country_count
            FROM dim_country;
        """)
    )

    print(result.fetchone())

(0,)


In [46]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                country_code,
                country_name,
                is_germany
            FROM dim_country
            WHERE country_code = 'DE';
        """)
    )

    print(result.fetchone())

None


In [47]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                COUNT(*) AS germany_flag_count
            FROM dim_country
            WHERE is_germany = TRUE;
        """)
    )

    print(result.fetchone())

(0,)


In [48]:
rows_inserted = country_sql.to_sql(
    name="dim_country",
    con=engine,
    if_exists="append",
    index=False
)

print("Rows inserted:", rows_inserted)

Rows inserted: 27


In [49]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                COUNT(*) AS country_count,
                SUM(is_germany) AS germany_flag_count
            FROM dim_country;
        """)
    )

    print(result.fetchone())

(27, Decimal('1'))


In [50]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                country_code,
                country_name,
                is_germany
            FROM dim_country
            WHERE country_code = 'DE';
        """)
    )

    print(result.fetchone())

('DE', 'Germany', 1)


In [51]:
source_market_sql = source_market_reference.copy()

source_market_sql = source_market_sql.rename(
    columns={
        "source_market": "source_market_name",
        "international_market": "is_eligible_market"
    }
)

print(source_market_sql.shape)
print(source_market_sql.columns.tolist())

display(
    source_market_sql.head(10)
)

(63, 4)
['source_market_code', 'source_market_name', 'market_type', 'is_eligible_market']


,source_market_code,source_market_name,market_type,is_eligible_market
0,AFR,Africa,regional_aggregate,False
1,AFR_OTH,Other African countries (aggregate changing ac...,regional_aggregate,False
2,AME,America,regional_aggregate,False
3,AME_C_S,Central and South America,regional_aggregate,False
4,AME_C_S_OTH,Other Central or South American countries,regional_aggregate,False
5,AME_N,Northern America,regional_aggregate,False
6,AME_N_OTH,Other Northern American countries,regional_aggregate,False
7,ASI,Asia,regional_aggregate,False
8,ASI_OTH,Other Asian countries (aggregate changing acco...,regional_aggregate,False
9,AT,Austria,individual_country,True


In [52]:
rows_inserted = source_market_sql.to_sql(
    name="dim_source_market",
    con=engine,
    if_exists="append",
    index=False
)

print("Rows inserted:", rows_inserted)

Rows inserted: 63


In [53]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                COUNT(*) AS total_markets,
                SUM(is_eligible_market) AS eligible_markets
            FROM dim_source_market;
        """)
    )

    print(result.fetchone())

(63, Decimal('41'))


In [54]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                market_type,
                COUNT(*) AS market_count,
                SUM(is_eligible_market) AS eligible_count
            FROM dim_source_market
            GROUP BY market_type
            ORDER BY market_type;
        """)
    )

    for row in result:
        print(row)

('combined_market', 1, Decimal('1'))
('domestic', 2, Decimal('0'))
('eu_aggregate', 4, Decimal('0'))
('individual_country', 40, Decimal('40'))
('regional_aggregate', 14, Decimal('0'))
('world_aggregate', 2, Decimal('0'))


In [55]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                source_market_code,
                source_market_name,
                market_type,
                is_eligible_market
            FROM dim_source_market
            WHERE source_market_code = 'CH_LI';
        """)
    )

    print(result.fetchone())

('CH_LI', 'Switzerland and Liechtenstein', 'combined_market', 1)


In [56]:
eu_annual = pd.read_csv(
    processed_dir
    / "eu27_tourism_benchmark_annual.csv"
)

print(eu_annual.shape)
print(eu_annual.columns.tolist())

display(
    eu_annual.head(10)
)

(123, 8)
['country_code', 'country', 'year', 'foreign_arrivals', 'foreign_nights', 'avg_length_of_stay', 'los_rank', 'countries_compared']


,country_code,country,year,foreign_arrivals,foreign_nights,avg_length_of_stay,los_rank,countries_compared
0,AT,Austria,2021,11098084.0,40034274.0,3.607314,8,22
1,AT,Austria,2022,22874333.0,79470210.0,3.474209,8,24
2,AT,Austria,2023,27212813.0,91023501.0,3.344877,9,24
3,AT,Austria,2024,28498633.0,94009113.0,3.298724,10,27
4,AT,Austria,2025,29764173.0,97020394.0,3.259637,9,26
5,BE,Belgium,2022,8220161.0,19377781.0,2.357348,18,24
6,BE,Belgium,2023,9279571.0,21594425.0,2.327093,19,24
7,BE,Belgium,2024,9638187.0,22279289.0,2.311564,20,27
8,BE,Belgium,2025,10127827.0,23360818.0,2.306597,18,26
9,BG,Bulgaria,2024,2994009.0,15058045.0,5.029392,3,27


In [57]:
print(
    "Years:",
    sorted(
        eu_annual["year"]
        .dropna()
        .unique()
    )
)

print(
    "\nRows by year:"
)

display(
    eu_annual
    .groupby("year")
    .size()
    .reset_index(name="row_count")
)

Years: [np.int64(2021), np.int64(2022), np.int64(2023), np.int64(2024), np.int64(2025)]

Rows by year:


,year,row_count
0,2021,22
1,2022,24
2,2023,24
3,2024,27
4,2025,26


In [58]:
display(
    eu_annual
    .isna()
    .sum()
    .to_frame("missing_values")
)

,missing_values
country_code,0
country,0
year,0
foreign_arrivals,0
foreign_nights,0
avg_length_of_stay,0
los_rank,0
countries_compared,0


In [59]:
eu_annual_sql = eu_annual[
    [
        "country_code",
        "year",
        "foreign_arrivals",
        "foreign_nights",
        "avg_length_of_stay",
        "los_rank",
        "countries_compared"
    ]
].copy()

print(eu_annual_sql.shape)
print(eu_annual_sql.columns.tolist())

display(eu_annual_sql.head())

(123, 7)
['country_code', 'year', 'foreign_arrivals', 'foreign_nights', 'avg_length_of_stay', 'los_rank', 'countries_compared']


,country_code,year,foreign_arrivals,foreign_nights,avg_length_of_stay,los_rank,countries_compared
0,AT,2021,11098084.0,40034274.0,3.607314,8,22
1,AT,2022,22874333.0,79470210.0,3.474209,8,24
2,AT,2023,27212813.0,91023501.0,3.344877,9,24
3,AT,2024,28498633.0,94009113.0,3.298724,10,27
4,AT,2025,29764173.0,97020394.0,3.259637,9,26


In [60]:
eu_annual_sql["year"] = eu_annual_sql["year"].astype(int)

eu_annual_sql["foreign_arrivals"] = (
    eu_annual_sql["foreign_arrivals"]
    .astype(int)
)

eu_annual_sql["foreign_nights"] = (
    eu_annual_sql["foreign_nights"]
    .astype(int)
)

eu_annual_sql["los_rank"] = (
    eu_annual_sql["los_rank"]
    .astype(int)
)

eu_annual_sql["countries_compared"] = (
    eu_annual_sql["countries_compared"]
    .astype(int)
)

print(eu_annual_sql.dtypes)

country_code              str
year                    int64
foreign_arrivals        int64
foreign_nights          int64
avg_length_of_stay    float64
los_rank                int64
countries_compared      int64
dtype: object


In [61]:
rows_inserted = eu_annual_sql.to_sql(
    name="fact_eu_tourism_annual",
    con=engine,
    if_exists="append",
    index=False
)

print("Rows inserted:", rows_inserted)

Rows inserted: 123


In [62]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                year,
                COUNT(*) AS country_count
            FROM fact_eu_tourism_annual
            GROUP BY year
            ORDER BY year;
        """)
    )

    for row in result:
        print(row)

(2021, 22)
(2022, 24)
(2023, 24)
(2024, 27)
(2025, 26)


In [63]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                year,
                foreign_arrivals,
                foreign_nights,
                avg_length_of_stay,
                los_rank,
                countries_compared
            FROM fact_eu_tourism_annual
            WHERE country_code = 'DE'
            ORDER BY year;
        """)
    )

    for row in result:
        print(row)

(2021, 11657007, 30731500, Decimal('2.6363'), 13, 22)
(2022, 28382919, 67616302, Decimal('2.3823'), 17, 24)
(2023, 34711761, 80378429, Decimal('2.3156'), 20, 24)
(2024, 37419799, 84792636, Decimal('2.2660'), 21, 27)
(2025, 37134349, 83075445, Decimal('2.2372'), 21, 26)


In [64]:
source_arrivals = pd.read_csv(
    clean_dir
    / "germany_source_market_arrivals_annual_clean.csv"
)

source_nights = pd.read_csv(
    clean_dir
    / "germany_source_market_nights_annual_clean.csv"
)

print("ARRIVALS")
print(source_arrivals.shape)
print(source_arrivals.columns.tolist())
display(source_arrivals.head())

print("\nNIGHTS")
print(source_nights.shape)
print(source_nights.columns.tolist())
display(source_nights.head())

ARRIVALS
(315, 10)
['destination_code', 'destination', 'source_market_code', 'source_market', 'year', 'foreign_arrivals', 'status', 'value_available', 'market_type', 'international_market']


,destination_code,destination,source_market_code,source_market,year,foreign_arrivals,status,value_available,market_type,international_market
0,DE,Germany,EUR,Europe,2021,92065343.0,NaN,True,regional_aggregate,False
1,DE,Germany,EUR,Europe,2022,152891215.0,NaN,True,regional_aggregate,False
2,DE,Germany,EUR,Europe,2023,170315670.0,NaN,True,regional_aggregate,False
3,DE,Germany,EUR,Europe,2024,175758213.0,NaN,True,regional_aggregate,False
4,DE,Germany,EUR,Europe,2025,NaN,NaN,False,regional_aggregate,False



NIGHTS
(315, 10)
['destination_code', 'destination', 'source_market_code', 'source_market', 'year', 'foreign_nights', 'status', 'value_available', 'market_type', 'international_market']


,destination_code,destination,source_market_code,source_market,year,foreign_nights,status,value_available,market_type,international_market
0,DE,Germany,EUR,Europe,2021,261386681.0,NaN,True,regional_aggregate,False
1,DE,Germany,EUR,Europe,2022,386328715.0,NaN,True,regional_aggregate,False
2,DE,Germany,EUR,Europe,2023,412829070.0,NaN,True,regional_aggregate,False
3,DE,Germany,EUR,Europe,2024,419240031.0,NaN,True,regional_aggregate,False
4,DE,Germany,EUR,Europe,2025,NaN,NaN,False,regional_aggregate,False


In [65]:
print("ARRIVALS YEARS")
print(sorted(source_arrivals["year"].dropna().unique()))

print("\nNIGHTS YEARS")
print(sorted(source_nights["year"].dropna().unique()))

print("\nARRIVALS MISSING VALUES")
display(
    source_arrivals
    .isna()
    .sum()
    .to_frame("missing_values")
)

print("\nNIGHTS MISSING VALUES")
display(
    source_nights
    .isna()
    .sum()
    .to_frame("missing_values")
)

ARRIVALS YEARS
[np.int64(2021), np.int64(2022), np.int64(2023), np.int64(2024), np.int64(2025)]

NIGHTS YEARS
[np.int64(2021), np.int64(2022), np.int64(2023), np.int64(2024), np.int64(2025)]

ARRIVALS MISSING VALUES


,missing_values
destination_code,0
destination,0
source_market_code,0
source_market,0
year,0
foreign_arrivals,67
status,315
value_available,0
market_type,0
international_market,0



NIGHTS MISSING VALUES


,missing_values
destination_code,0
destination,0
source_market_code,0
source_market,0
year,0
foreign_nights,67
status,315
value_available,0
market_type,0
international_market,0


In [66]:
arrivals_keys = set(
    zip(
        source_arrivals["source_market_code"],
        source_arrivals["year"]
    )
)

nights_keys = set(
    zip(
        source_nights["source_market_code"],
        source_nights["year"]
    )
)

print(
    "Arrival keys:",
    len(arrivals_keys)
)

print(
    "Night keys:",
    len(nights_keys)
)

print(
    "Only in arrivals:",
    len(arrivals_keys - nights_keys)
)

print(
    "Only in nights:",
    len(nights_keys - arrivals_keys)
)

Arrival keys: 315
Night keys: 315
Only in arrivals: 0
Only in nights: 0


In [67]:
source_market_annual = source_arrivals[
    [
        "source_market_code",
        "year",
        "foreign_arrivals",
        "value_available"
    ]
].rename(
    columns={
        "value_available": "arrivals_available"
    }
).merge(
    source_nights[
        [
            "source_market_code",
            "year",
            "foreign_nights",
            "value_available"
        ]
    ].rename(
        columns={
            "value_available": "nights_available"
        }
    ),
    on=[
        "source_market_code",
        "year"
    ],
    how="inner",
    validate="one_to_one"
)

print(source_market_annual.shape)

display(
    source_market_annual.head(10)
)

(315, 6)


,source_market_code,year,foreign_arrivals,arrivals_available,foreign_nights,nights_available
0,EUR,2021,92065343.0,True,261386681.0,True
1,EUR,2022,152891215.0,True,386328715.0,True
2,EUR,2023,170315670.0,True,412829070.0,True
3,EUR,2024,175758213.0,True,419240031.0,True
4,EUR,2025,NaN,False,NaN,False
5,BE,2021,621993.0,True,1394320.0,True
6,BE,2022,1216635.0,True,2617268.0,True
7,BE,2023,1396964.0,True,2939910.0,True
8,BE,2024,1465552.0,True,3044527.0,True
9,BE,2025,NaN,False,NaN,False


In [68]:
display(
    source_market_annual
    .groupby(
        [
            "arrivals_available",
            "nights_available"
        ]
    )
    .size()
    .reset_index(name="row_count")
)

,arrivals_available,nights_available,row_count
0,False,False,67
1,True,True,248


In [69]:
print(
    "Missing arrivals:",
    source_market_annual["foreign_arrivals"].isna().sum()
)

print(
    "Missing nights:",
    source_market_annual["foreign_nights"].isna().sum()
)

print(
    "Both measures available:",
    (
        source_market_annual["arrivals_available"]
        & source_market_annual["nights_available"]
    ).sum()
)

Missing arrivals: 67
Missing nights: 67
Both measures available: 248


In [70]:
eligible_market_codes = (
    source_market_reference
    .loc[
        source_market_reference["international_market"],
        "source_market_code"
    ]
    .tolist()
)

print(
    "Eligible market codes:",
    len(eligible_market_codes)
)

Eligible market codes: 41


In [71]:
germany_source_market_sql = (
    source_market_annual[
        source_market_annual["source_market_code"]
        .isin(eligible_market_codes)
    ]
    .loc[
        lambda df:
        df["arrivals_available"]
        & df["nights_available"]
    ]
    .copy()
)

germany_source_market_sql[
    "avg_length_of_stay"
] = (
    germany_source_market_sql["foreign_nights"]
    / germany_source_market_sql["foreign_arrivals"]
)

germany_source_market_sql = germany_source_market_sql[
    [
        "source_market_code",
        "year",
        "foreign_arrivals",
        "foreign_nights",
        "avg_length_of_stay"
    ]
]

print(germany_source_market_sql.shape)

display(
    germany_source_market_sql.head(10)
)

(164, 5)


,source_market_code,year,foreign_arrivals,foreign_nights,avg_length_of_stay
5,BE,2021,621993.0,1394320.0,2.241697
6,BE,2022,1216635.0,2617268.0,2.151235
7,BE,2023,1396964.0,2939910.0,2.104499
8,BE,2024,1465552.0,3044527.0,2.077393
10,BG,2021,57481.0,275345.0,4.790192
11,BG,2022,104700.0,400930.0,3.829322
12,BG,2023,137512.0,461877.0,3.358812
13,BG,2024,147766.0,511086.0,3.458752
15,CZ,2021,227946.0,672391.0,2.949782
16,CZ,2022,503872.0,1197873.0,2.377336


In [72]:
germany_source_market_sql["year"] = (
    germany_source_market_sql["year"]
    .astype(int)
)

germany_source_market_sql["foreign_arrivals"] = (
    germany_source_market_sql["foreign_arrivals"]
    .astype(int)
)

germany_source_market_sql["foreign_nights"] = (
    germany_source_market_sql["foreign_nights"]
    .astype(int)
)

In [73]:
display(
    germany_source_market_sql
    .groupby("year")
    .size()
    .reset_index(name="market_count")
)

,year,market_count
0,2021,41
1,2022,41
2,2023,41
3,2024,41


In [74]:
print(
    "Unique eligible markets represented:",
    germany_source_market_sql[
        "source_market_code"
    ].nunique()
)

print(
    "Duplicate market-year rows:",
    germany_source_market_sql.duplicated(
        subset=[
            "source_market_code",
            "year"
        ]
    ).sum()
)

Unique eligible markets represented: 41
Duplicate market-year rows: 0


In [75]:
rows_inserted = germany_source_market_sql.to_sql(
    name="fact_germany_source_market",
    con=engine,
    if_exists="append",
    index=False
)

print("Rows inserted:", rows_inserted)

Rows inserted: 164


In [76]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                year,
                COUNT(*) AS market_count
            FROM fact_germany_source_market
            GROUP BY year
            ORDER BY year;
        """)
    )

    for row in result:
        print(row)

(2021, 41)
(2022, 41)
(2023, 41)
(2024, 41)


In [77]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                COUNT(*) AS markets,
                SUM(foreign_arrivals) AS published_market_arrivals,
                SUM(foreign_nights) AS published_market_nights
            FROM fact_germany_source_market
            WHERE year = 2024;
        """)
    )

    print(result.fetchone())

(41, Decimal('33318217'), Decimal('75112974'))


In [78]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                source_market_code,
                year,
                foreign_arrivals,
                foreign_nights,
                avg_length_of_stay
            FROM fact_germany_source_market
            WHERE source_market_code = 'NL'
              AND year = 2024;
        """)
    )

    print(result.fetchone())

('NL', 2024, 5250119, 11867612, Decimal('2.2604'))


In [79]:
germany_monthly_arrivals = pd.read_csv(
    clean_dir
    / "germany_foreign_arrivals_monthly_clean.csv"
)

germany_monthly_nights = pd.read_csv(
    clean_dir
    / "germany_foreign_nights_monthly_clean.csv"
)

print("ARRIVALS")
print(germany_monthly_arrivals.shape)
print(germany_monthly_arrivals.columns.tolist())
display(germany_monthly_arrivals.head())

print("\nNIGHTS")
print(germany_monthly_nights.shape)
print(germany_monthly_nights.columns.tolist())
display(germany_monthly_nights.head())

ARRIVALS
(67, 11)
['country_code', 'country', 'period', 'foreign_arrivals', 'status', 'year', 'month', 'value_available', 'date', 'months_available', 'complete_12_month_year']


,country_code,country,period,foreign_arrivals,status,year,month,value_available,date,months_available,complete_12_month_year
0,DE,Germany,2021-01,169984.0,NaN,2021,1,True,2021-01-01,12,True
1,DE,Germany,2021-02,186270.0,NaN,2021,2,True,2021-02-01,12,True
2,DE,Germany,2021-03,230216.0,NaN,2021,3,True,2021-03-01,12,True
3,DE,Germany,2021-04,219128.0,NaN,2021,4,True,2021-04-01,12,True
4,DE,Germany,2021-05,291267.0,NaN,2021,5,True,2021-05-01,12,True



NIGHTS
(67, 11)
['country_code', 'country', 'period', 'foreign_nights', 'status', 'year', 'month', 'value_available', 'date', 'months_available', 'complete_12_month_year']


,country_code,country,period,foreign_nights,status,year,month,value_available,date,months_available,complete_12_month_year
0,DE,Germany,2021-01,692238.0,NaN,2021,1,True,2021-01-01,12,True
1,DE,Germany,2021-02,771308.0,NaN,2021,2,True,2021-02-01,12,True
2,DE,Germany,2021-03,930822.0,NaN,2021,3,True,2021-03-01,12,True
3,DE,Germany,2021-04,878134.0,NaN,2021,4,True,2021-04-01,12,True
4,DE,Germany,2021-05,1029929.0,NaN,2021,5,True,2021-05-01,12,True


In [80]:
print("ARRIVALS YEARS")
print(
    sorted(
        germany_monthly_arrivals["year"]
        .dropna()
        .unique()
    )
)

print("\nNIGHTS YEARS")
print(
    sorted(
        germany_monthly_nights["year"]
        .dropna()
        .unique()
    )
)

ARRIVALS YEARS
[np.int64(2021), np.int64(2022), np.int64(2023), np.int64(2024), np.int64(2025), np.int64(2026)]

NIGHTS YEARS
[np.int64(2021), np.int64(2022), np.int64(2023), np.int64(2024), np.int64(2025), np.int64(2026)]


In [81]:
print("ARRIVALS MISSING")
display(
    germany_monthly_arrivals
    .isna()
    .sum()
    .to_frame("missing_values")
)

print("\nNIGHTS MISSING")
display(
    germany_monthly_nights
    .isna()
    .sum()
    .to_frame("missing_values")
)

ARRIVALS MISSING


,missing_values
country_code,0
country,0
period,0
foreign_arrivals,1
status,67
year,0
month,0
value_available,0
date,0
months_available,0



NIGHTS MISSING


,missing_values
country_code,0
country,0
period,0
foreign_nights,1
status,67
year,0
month,0
value_available,0
date,0
months_available,0


In [82]:
arrivals_month_keys = set(
    zip(
        germany_monthly_arrivals["year"],
        germany_monthly_arrivals["month"]
    )
)

nights_month_keys = set(
    zip(
        germany_monthly_nights["year"],
        germany_monthly_nights["month"]
    )
)

print(
    "Arrival month keys:",
    len(arrivals_month_keys)
)

print(
    "Night month keys:",
    len(nights_month_keys)
)

print(
    "Only in arrivals:",
    len(arrivals_month_keys - nights_month_keys)
)

print(
    "Only in nights:",
    len(nights_month_keys - arrivals_month_keys)
)

Arrival month keys: 67
Night month keys: 67
Only in arrivals: 0
Only in nights: 0


In [83]:
monthly_availability = (
    germany_monthly_arrivals[
        [
            "year",
            "month",
            "date",
            "foreign_arrivals",
            "value_available"
        ]
    ]
    .rename(
        columns={
            "value_available": "arrivals_available"
        }
    )
    .merge(
        germany_monthly_nights[
            [
                "year",
                "month",
                "date",
                "foreign_nights",
                "value_available"
            ]
        ].rename(
            columns={
                "value_available": "nights_available"
            }
        ),
        on=[
            "year",
            "month",
            "date"
        ],
        how="inner",
        validate="one_to_one"
    )
)

display(
    monthly_availability
    .groupby(
        [
            "arrivals_available",
            "nights_available"
        ]
    )
    .size()
    .reset_index(name="row_count")
)

,arrivals_available,nights_available,row_count
0,False,False,1
1,True,True,66


In [84]:
display(
    monthly_availability[
        monthly_availability["year"] == 2026
    ]
)

,year,month,date,foreign_arrivals,arrivals_available,foreign_nights,nights_available
60,2026,1,2026-01-01,1938956.0,True,4277203.0,True
61,2026,2,2026-02-01,2271693.0,True,5078278.0,True
62,2026,3,2026-03-01,2341823.0,True,5197078.0,True
63,2026,4,2026-04-01,2934984.0,True,6596278.0,True
64,2026,5,2026-05-01,3353419.0,True,7446648.0,True
65,2026,6,2026-06-01,3505228.0,True,7653746.0,True
66,2026,7,2026-07-01,NaN,False,NaN,False


In [85]:
monthly_availability["both_available"] = (
    monthly_availability["arrivals_available"]
    & monthly_availability["nights_available"]
)

display(
    monthly_availability
    .groupby("year")
    .agg(
        periods_present=("month", "count"),
        observed_periods=("both_available", "sum")
    )
    .reset_index()
)

,year,periods_present,observed_periods
0,2021,12,12
1,2022,12,12
2,2023,12,12
3,2024,12,12
4,2025,12,12
5,2026,7,6


In [86]:
germany_monthly_sql = (
    monthly_availability[
        monthly_availability["both_available"]
    ]
    .copy()
)

germany_monthly_sql[
    "avg_length_of_stay"
] = (
    germany_monthly_sql["foreign_nights"]
    / germany_monthly_sql["foreign_arrivals"]
)

germany_monthly_sql = germany_monthly_sql.rename(
    columns={
        "date": "period_date"
    }
)

germany_monthly_sql = germany_monthly_sql[
    [
        "period_date",
        "year",
        "month",
        "foreign_arrivals",
        "foreign_nights",
        "avg_length_of_stay"
    ]
].copy()

germany_monthly_sql["period_date"] = pd.to_datetime(
    germany_monthly_sql["period_date"]
)

germany_monthly_sql["year"] = (
    germany_monthly_sql["year"]
    .astype(int)
)

germany_monthly_sql["month"] = (
    germany_monthly_sql["month"]
    .astype(int)
)

germany_monthly_sql["foreign_arrivals"] = (
    germany_monthly_sql["foreign_arrivals"]
    .astype(int)
)

germany_monthly_sql["foreign_nights"] = (
    germany_monthly_sql["foreign_nights"]
    .astype(int)
)

print(germany_monthly_sql.shape)
print(germany_monthly_sql.dtypes)

display(
    germany_monthly_sql.tail(10)
)

(66, 6)
period_date           datetime64[us]
year                           int64
month                          int64
foreign_arrivals               int64
foreign_nights                 int64
avg_length_of_stay           float64
dtype: object


,period_date,year,month,foreign_arrivals,foreign_nights,avg_length_of_stay
56,2025-09-01,2025,9,3451143,7674317,2.223703
57,2025-10-01,2025,10,3171850,7195093,2.268422
58,2025-11-01,2025,11,2550385,5590732,2.192113
59,2025-12-01,2025,12,3045185,6582972,2.161764
60,2026-01-01,2026,1,1938956,4277203,2.205931
61,2026-02-01,2026,2,2271693,5078278,2.235460
62,2026-03-01,2026,3,2341823,5197078,2.219245
63,2026-04-01,2026,4,2934984,6596278,2.247466
64,2026-05-01,2026,5,3353419,7446648,2.220614
65,2026-06-01,2026,6,3505228,7653746,2.183523


In [87]:
print(
    "First observed month:",
    germany_monthly_sql["period_date"].min()
)

print(
    "Last observed month:",
    germany_monthly_sql["period_date"].max()
)

print(
    "Duplicate dates:",
    germany_monthly_sql["period_date"].duplicated().sum()
)

print(
    "Missing values:",
    germany_monthly_sql.isna().sum().sum()
)

First observed month: 2021-01-01 00:00:00
Last observed month: 2026-06-01 00:00:00
Duplicate dates: 0
Missing values: 0


In [88]:
display(
    germany_monthly_sql[
        germany_monthly_sql["year"] == 2026
    ]
    .agg(
        {
            "foreign_arrivals": "sum",
            "foreign_nights": "sum"
        }
    )
)

foreign_arrivals    16346103
foreign_nights      36249231
dtype: int64

In [89]:
rows_inserted = germany_monthly_sql.to_sql(
    name="fact_germany_tourism_monthly",
    con=engine,
    if_exists="append",
    index=False
)

print("Rows inserted:", rows_inserted)

Rows inserted: 66


In [90]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                year,
                COUNT(*) AS observed_months
            FROM fact_germany_tourism_monthly
            GROUP BY year
            ORDER BY year;
        """)
    )

    for row in result:
        print(row)

(2021, 12)
(2022, 12)
(2023, 12)
(2024, 12)
(2025, 12)
(2026, 6)


In [91]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                COUNT(*) AS months,
                SUM(foreign_arrivals) AS foreign_arrivals,
                SUM(foreign_nights) AS foreign_nights,
                ROUND(
                    SUM(foreign_nights)
                    / SUM(foreign_arrivals),
                    3
                ) AS h1_avg_length_of_stay
            FROM fact_germany_tourism_monthly
            WHERE year = 2026;
        """)
    )

    print(result.fetchone())

(6, Decimal('16346103'), Decimal('36249231'), Decimal('2.218'))


In [92]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT MAX(period_date)
            FROM fact_germany_tourism_monthly;
        """)
    )

    print(result.fetchone())

(datetime.date(2026, 6, 1),)


In [93]:
economic_files = sorted(
    [
        file.name
        for file in processed_dir.glob("*")
        if "economic" in file.name.lower()
        or "bop" in file.name.lower()
        or "travel" in file.name.lower()
    ]
)

economic_files

['germany_strategic_economic_context.csv',
 'germany_travel_bop_identity_check.csv',
 'germany_travel_bop_summary_2021_2025.csv',
 'germany_travel_receipts_by_partner_2021_2025.csv',
 'germany_travel_receipts_partner_growth_2021_2025.csv']

In [94]:
germany_economic = pd.read_csv(
    processed_dir
    / "germany_travel_bop_summary_2021_2025.csv"
)

print(germany_economic.shape)
print(germany_economic.columns.tolist())

display(germany_economic)

(5, 12)
['year', 'travel_balance_mio_eur', 'travel_receipts_mio_eur', 'travel_expenditure_mio_eur', 'receipts_expenditure_ratio', 'receipts_cover_expenditure_pct', 'is_provisional', 'receipts_growth_pct', 'expenditure_growth_pct', 'balance_change_mio_eur', 'receipts_change_mio_eur', 'expenditure_change_mio_eur']


,year,travel_balance_mio_eur,travel_receipts_mio_eur,travel_expenditure_mio_eur,receipts_expenditure_ratio,receipts_cover_expenditure_pct,is_provisional,receipts_growth_pct,expenditure_growth_pct,balance_change_mio_eur,receipts_change_mio_eur,expenditure_change_mio_eur
0,2021,-24300.0,18827.0,43126.0,0.436558,43.655799,False,NaN,NaN,NaN,NaN,NaN
1,2022,-54762.0,30257.0,85019.0,0.355885,35.588516,False,60.710681,97.140936,-30462.0,11430.0,41893.0
2,2023,-71650.0,34992.0,106642.0,0.328126,32.812588,False,15.649271,25.433138,-16888.0,4735.0,21623.0
3,2024,-69767.0,37055.0,106822.0,0.346885,34.688547,False,5.895633,0.168789,1883.0,2063.0,180.0
4,2025,-76822.0,37772.0,114594.0,0.329616,32.961586,True,1.934962,7.275655,-7055.0,717.0,7772.0


In [95]:
print(germany_economic.dtypes)

display(
    germany_economic
    .isna()
    .sum()
    .to_frame("missing_values")
)

year                                int64
travel_balance_mio_eur            float64
travel_receipts_mio_eur           float64
travel_expenditure_mio_eur        float64
receipts_expenditure_ratio        float64
receipts_cover_expenditure_pct    float64
is_provisional                       bool
receipts_growth_pct               float64
expenditure_growth_pct            float64
balance_change_mio_eur            float64
receipts_change_mio_eur           float64
expenditure_change_mio_eur        float64
dtype: object


,missing_values
year,0
travel_balance_mio_eur,0
travel_receipts_mio_eur,0
travel_expenditure_mio_eur,0
receipts_expenditure_ratio,0
receipts_cover_expenditure_pct,0
is_provisional,0
receipts_growth_pct,1
expenditure_growth_pct,1
balance_change_mio_eur,1


In [96]:
print(
    "Years:",
    sorted(
        germany_economic["year"]
        .dropna()
        .unique()
    )
)

Years: [np.int64(2021), np.int64(2022), np.int64(2023), np.int64(2024), np.int64(2025)]


In [97]:
germany_economic_sql = germany_economic[
    [
        "year",
        "travel_receipts_mio_eur",
        "travel_expenditure_mio_eur",
        "travel_balance_mio_eur",
        "receipts_cover_expenditure_pct",
        "receipts_growth_pct",
        "expenditure_growth_pct",
        "is_provisional"
    ]
].copy()

print(germany_economic_sql.shape)
print(germany_economic_sql.dtypes)

display(germany_economic_sql)

(5, 8)
year                                int64
travel_receipts_mio_eur           float64
travel_expenditure_mio_eur        float64
travel_balance_mio_eur            float64
receipts_cover_expenditure_pct    float64
receipts_growth_pct               float64
expenditure_growth_pct            float64
is_provisional                       bool
dtype: object


,year,travel_receipts_mio_eur,travel_expenditure_mio_eur,travel_balance_mio_eur,receipts_cover_expenditure_pct,receipts_growth_pct,expenditure_growth_pct,is_provisional
0,2021,18827.0,43126.0,-24300.0,43.655799,NaN,NaN,False
1,2022,30257.0,85019.0,-54762.0,35.588516,60.710681,97.140936,False
2,2023,34992.0,106642.0,-71650.0,32.812588,15.649271,25.433138,False
3,2024,37055.0,106822.0,-69767.0,34.688547,5.895633,0.168789,False
4,2025,37772.0,114594.0,-76822.0,32.961586,1.934962,7.275655,True


In [98]:
rows_inserted = germany_economic_sql.to_sql(
    name="fact_germany_travel_economic",
    con=engine,
    if_exists="append",
    index=False
)

print("Rows inserted:", rows_inserted)

Rows inserted: 5


In [99]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                year,
                travel_receipts_mio_eur,
                travel_expenditure_mio_eur,
                travel_balance_mio_eur,
                receipts_cover_expenditure_pct,
                receipts_growth_pct,
                expenditure_growth_pct,
                is_provisional
            FROM fact_germany_travel_economic
            ORDER BY year;
        """)
    )

    for row in result:
        print(row)

(2021, Decimal('18827.00'), Decimal('43126.00'), Decimal('-24300.00'), Decimal('43.6558'), None, None, 0)
(2022, Decimal('30257.00'), Decimal('85019.00'), Decimal('-54762.00'), Decimal('35.5885'), Decimal('60.7107'), Decimal('97.1409'), 0)
(2023, Decimal('34992.00'), Decimal('106642.00'), Decimal('-71650.00'), Decimal('32.8126'), Decimal('15.6493'), Decimal('25.4331'), 0)
(2024, Decimal('37055.00'), Decimal('106822.00'), Decimal('-69767.00'), Decimal('34.6885'), Decimal('5.8956'), Decimal('0.1688'), 0)
(2025, Decimal('37772.00'), Decimal('114594.00'), Decimal('-76822.00'), Decimal('32.9616'), Decimal('1.9350'), Decimal('7.2757'), 1)


In [100]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                COUNT(*) AS total_years,
                SUM(is_provisional) AS provisional_years,
                SUM(
                    CASE
                        WHEN receipts_growth_pct IS NULL
                        THEN 1
                        ELSE 0
                    END
                ) AS null_receipts_growth,
                SUM(
                    CASE
                        WHEN expenditure_growth_pct IS NULL
                        THEN 1
                        ELSE 0
                    END
                ) AS null_expenditure_growth
            FROM fact_germany_travel_economic;
        """)
    )

    print(result.fetchone())

(5, Decimal('1'), Decimal('1'), Decimal('1'))


In [101]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                year,
                is_provisional
            FROM fact_germany_travel_economic
            WHERE is_provisional = TRUE;
        """)
    )

    print(result.fetchone())

(2025, 1)


In [102]:
strategy_files = sorted(
    [
        file.name
        for file in processed_dir.glob("*")
        if "strategy" in file.name.lower()
        or "strategic" in file.name.lower()
        or "source_market" in file.name.lower()
    ]
)

strategy_files

['germany_integrated_strategic_diagnosis.csv',
 'germany_major_source_market_strategy_2024.csv',
 'germany_source_market_action_matrix.csv',
 'germany_source_market_analysis_2024.csv',
 'germany_source_market_coverage_2021_2024.csv',
 'germany_source_market_strategic_roles_2024.csv',
 'germany_source_market_threshold_sensitivity.csv',
 'germany_stay_duration_opportunity_strategy.csv',
 'germany_strategic_economic_context.csv',
 'germany_strategic_evidence_summary.csv',
 'germany_strategic_priority_matrix.csv',
 'germany_strategic_recommendations.csv',
 'germany_strategic_role_summary.csv',
 'germany_strategy_implementation_framework.csv']

In [103]:
source_strategy = pd.read_csv(
    processed_dir
    / "germany_major_source_market_strategy_2024.csv"
)

print(source_strategy.shape)
print(source_strategy.columns.tolist())

display(source_strategy)

(16, 16)
['source_market_code', 'source_market', 'market_type', 'foreign_nights', 'foreign_arrivals', 'germany_foreign_nights_share_pct', 'published_market_share_pct', 'growth_2023_2024_pct', 'absolute_growth_nights', 'avg_length_of_stay', 'high_scale', 'high_growth', 'longer_stay', 'opportunity_score', 'strategic_role', 'stay_extension_priority']


,source_market_code,source_market,market_type,foreign_nights,foreign_arrivals,germany_foreign_nights_share_pct,published_market_share_pct,growth_2023_2024_pct,absolute_growth_nights,avg_length_of_stay,high_scale,high_growth,longer_stay,opportunity_score,strategic_role,stay_extension_priority
0,US,United States,individual_country,7223819.0,3222927.0,8.519394,9.617272,10.750402,701207.0,2.241385,True,True,True,3,Core Growth Market,Lower
1,UK,United Kingdom,individual_country,5274041.0,2373132.0,6.219928,7.021478,10.880105,517515.0,2.222397,True,True,True,3,Core Growth Market,Lower
2,NL,Netherlands,individual_country,11867612.0,5250119.0,13.996041,15.799683,3.659706,418986.0,2.260446,True,False,True,2,Core Retention Market,Lower
3,AT,Austria,individual_country,4360393.0,2131281.0,5.142419,5.805113,4.927288,204760.0,2.045902,True,True,False,2,Stay Extension Opportunity,High
4,PL,Poland,individual_country,4167089.0,1471375.0,4.914447,5.547762,0.155626,6475.0,2.832105,True,False,True,2,Core Retention Market,Lower
5,FR,France,individual_country,3680328.0,1849028.0,4.340386,4.899723,4.836640,169792.0,1.990412,True,True,False,2,Stay Extension Opportunity,High
6,IT,Italy,individual_country,3497676.0,1577424.0,4.124976,4.656554,4.605835,154004.0,2.217334,True,True,False,2,Stay Extension Opportunity,High
7,ES,Spain,individual_country,2720461.0,1123109.0,3.208369,3.621826,7.653031,193397.0,2.422259,False,True,True,2,Emerging Growth Market,Lower
8,TR,Türkiye,individual_country,1225934.0,488285.0,1.445802,1.632120,17.945975,186531.0,2.510694,False,True,True,2,Emerging Growth Market,Lower
9,CH_LI,Switzerland and Liechtenstein,combined_market,6897129.0,3315820.0,8.134113,9.182340,2.414773,162623.0,2.080067,True,False,False,1,Monitor / Maintain,High


In [104]:
print(source_strategy.dtypes)

display(
    source_strategy
    .isna()
    .sum()
    .to_frame("missing_values")
)

source_market_code                      str
source_market                           str
market_type                             str
foreign_nights                      float64
foreign_arrivals                    float64
germany_foreign_nights_share_pct    float64
published_market_share_pct          float64
growth_2023_2024_pct                float64
absolute_growth_nights              float64
avg_length_of_stay                  float64
high_scale                             bool
high_growth                            bool
longer_stay                            bool
opportunity_score                     int64
strategic_role                          str
stay_extension_priority                 str
dtype: object


,missing_values
source_market_code,0
source_market,0
market_type,0
foreign_nights,0
foreign_arrivals,0
germany_foreign_nights_share_pct,0
published_market_share_pct,0
growth_2023_2024_pct,0
absolute_growth_nights,0
avg_length_of_stay,0


In [105]:
source_action = pd.read_csv(
    processed_dir
    / "germany_source_market_action_matrix.csv"
)

print(source_action.shape)
print(source_action.columns.tolist())

display(source_action)

(16, 10)
['source_market_code', 'source_market', 'foreign_nights', 'germany_foreign_nights_share_pct', 'growth_2023_2024_pct', 'absolute_growth_nights', 'avg_length_of_stay', 'strategic_role', 'stay_extension_priority', 'recommended_action']


,source_market_code,source_market,foreign_nights,germany_foreign_nights_share_pct,growth_2023_2024_pct,absolute_growth_nights,avg_length_of_stay,strategic_role,stay_extension_priority,recommended_action
0,US,United States,7223819.0,8.519394,10.750402,701207.0,2.241385,Core Growth Market,Lower,Protect growth while strengthening repeat visi...
1,UK,United Kingdom,5274041.0,6.219928,10.880105,517515.0,2.222397,Core Growth Market,Lower,Protect growth while strengthening repeat visi...
2,NL,Netherlands,11867612.0,13.996041,3.659706,418986.0,2.260446,Core Retention Market,Lower,Defend market share and maintain strong visito...
3,AT,Austria,4360393.0,5.142419,4.927288,204760.0,2.045902,Stay Extension Opportunity,High,"Prioritize longer-stay products, multi-destina..."
4,PL,Poland,4167089.0,4.914447,0.155626,6475.0,2.832105,Core Retention Market,Lower,Defend market share and maintain strong visito...
5,FR,France,3680328.0,4.340386,4.836640,169792.0,1.990412,Stay Extension Opportunity,High,"Prioritize longer-stay products, multi-destina..."
6,IT,Italy,3497676.0,4.124976,4.605835,154004.0,2.217334,Stay Extension Opportunity,High,"Prioritize longer-stay products, multi-destina..."
7,ES,Spain,2720461.0,3.208369,7.653031,193397.0,2.422259,Emerging Growth Market,Lower,Support growth selectively and monitor whether...
8,TR,Türkiye,1225934.0,1.445802,17.945975,186531.0,2.510694,Emerging Growth Market,Lower,Support growth selectively and monitor whether...
9,CH_LI,Switzerland and Liechtenstein,6897129.0,8.134113,2.414773,162623.0,2.080067,Monitor / Maintain,High,Maintain presence and monitor for changes in s...


In [106]:
print(source_action.dtypes)

display(
    source_action
    .isna()
    .sum()
    .to_frame("missing_values")
)

source_market_code                      str
source_market                           str
foreign_nights                      float64
germany_foreign_nights_share_pct    float64
growth_2023_2024_pct                float64
absolute_growth_nights              float64
avg_length_of_stay                  float64
strategic_role                          str
stay_extension_priority                 str
recommended_action                      str
dtype: object


,missing_values
source_market_code,0
source_market,0
foreign_nights,0
germany_foreign_nights_share_pct,0
growth_2023_2024_pct,0
absolute_growth_nights,0
avg_length_of_stay,0
strategic_role,0
stay_extension_priority,0
recommended_action,0


In [107]:
source_market_strategy_sql = source_action.copy()

source_market_strategy_sql["analysis_year"] = 2024

source_market_strategy_sql["foreign_nights"] = (
    source_market_strategy_sql["foreign_nights"]
    .astype(int)
)

source_market_strategy_sql["absolute_growth_nights"] = (
    source_market_strategy_sql["absolute_growth_nights"]
    .astype(int)
)

source_market_strategy_sql = source_market_strategy_sql[
    [
        "source_market_code",
        "analysis_year",
        "foreign_nights",
        "germany_foreign_nights_share_pct",
        "growth_2023_2024_pct",
        "absolute_growth_nights",
        "avg_length_of_stay",
        "strategic_role",
        "stay_extension_priority",
        "recommended_action"
    ]
].copy()

print(source_market_strategy_sql.shape)
print(source_market_strategy_sql.dtypes)

display(source_market_strategy_sql)

(16, 10)
source_market_code                      str
analysis_year                         int64
foreign_nights                        int64
germany_foreign_nights_share_pct    float64
growth_2023_2024_pct                float64
absolute_growth_nights                int64
avg_length_of_stay                  float64
strategic_role                          str
stay_extension_priority                 str
recommended_action                      str
dtype: object


,source_market_code,analysis_year,foreign_nights,germany_foreign_nights_share_pct,growth_2023_2024_pct,absolute_growth_nights,avg_length_of_stay,strategic_role,stay_extension_priority,recommended_action
0,US,2024,7223819,8.519394,10.750402,701207,2.241385,Core Growth Market,Lower,Protect growth while strengthening repeat visi...
1,UK,2024,5274041,6.219928,10.880105,517515,2.222397,Core Growth Market,Lower,Protect growth while strengthening repeat visi...
2,NL,2024,11867612,13.996041,3.659706,418986,2.260446,Core Retention Market,Lower,Defend market share and maintain strong visito...
3,AT,2024,4360393,5.142419,4.927288,204760,2.045902,Stay Extension Opportunity,High,"Prioritize longer-stay products, multi-destina..."
4,PL,2024,4167089,4.914447,0.155626,6475,2.832105,Core Retention Market,Lower,Defend market share and maintain strong visito...
5,FR,2024,3680328,4.340386,4.836640,169792,1.990412,Stay Extension Opportunity,High,"Prioritize longer-stay products, multi-destina..."
6,IT,2024,3497676,4.124976,4.605835,154004,2.217334,Stay Extension Opportunity,High,"Prioritize longer-stay products, multi-destina..."
7,ES,2024,2720461,3.208369,7.653031,193397,2.422259,Emerging Growth Market,Lower,Support growth selectively and monitor whether...
8,TR,2024,1225934,1.445802,17.945975,186531,2.510694,Emerging Growth Market,Lower,Support growth selectively and monitor whether...
9,CH_LI,2024,6897129,8.134113,2.414773,162623,2.080067,Monitor / Maintain,High,Maintain presence and monitor for changes in s...


In [108]:
with engine.connect() as connection:
    source_codes_db = pd.read_sql(
        """
        SELECT source_market_code
        FROM dim_source_market;
        """,
        connection
    )

missing_codes = set(
    source_market_strategy_sql["source_market_code"]
) - set(
    source_codes_db["source_market_code"]
)

print("Missing source-market codes:", missing_codes)

Missing source-market codes: set()


In [109]:
print(
    "Duplicate market-year rows:",
    source_market_strategy_sql[
        ["source_market_code", "analysis_year"]
    ]
    .duplicated()
    .sum()
)

Duplicate market-year rows: 0


In [110]:
rows_inserted = source_market_strategy_sql.to_sql(
    name="analysis_source_market_strategy",
    con=engine,
    if_exists="append",
    index=False
)

print("Rows inserted:", rows_inserted)

Rows inserted: 16


In [111]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                analysis_year,
                COUNT(*) AS markets,
                SUM(foreign_nights) AS foreign_nights,
                SUM(absolute_growth_nights) AS absolute_growth_nights
            FROM analysis_source_market_strategy
            GROUP BY analysis_year;
        """)
    )

    print(result.fetchone())

(2024, 16, Decimal('63163500'), Decimal('3517249'))


In [112]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                strategic_role,
                COUNT(*) AS markets
            FROM analysis_source_market_strategy
            GROUP BY strategic_role
            ORDER BY markets DESC, strategic_role;
        """)
    )

    for row in result:
        print(row)

('Monitor / Maintain', 4)
('Stay Extension Opportunity', 3)
('Core Growth Market', 2)
('Core Retention Market', 2)
('Emerging Growth Market', 2)
('Long-Stay Niche', 2)
('Growth Market - Shorter Stay', 1)


In [113]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                stay_extension_priority,
                COUNT(*) AS markets
            FROM analysis_source_market_strategy
            GROUP BY stay_extension_priority
            ORDER BY stay_extension_priority;
        """)
    )

    for row in result:
        print(row)

('Emerging', 1)
('High', 4)
('Lower', 11)


In [114]:
los_files = sorted(
    [
        file.name
        for file in processed_dir.glob("*")
        if "los" in file.name.lower()
        or "scenario" in file.name.lower()
        or "counterfactual" in file.name.lower()
        or "benchmark" in file.name.lower()
    ]
)

los_files

['eu27_2026_ytd_arrivals_growth_benchmark.csv',
 'eu27_2026_ytd_los_benchmark.csv',
 'eu27_2026_ytd_nights_benchmark.csv',
 'eu27_tourism_benchmark_2024.csv',
 'eu27_tourism_benchmark_annual.csv',
 'germany_2026_baseline_scenario_summary.csv',
 'germany_2026_ytd_los_gap.csv',
 'germany_eu_2026_ytd_los_summary.csv',
 'germany_eu_los_benchmark_2021_2025.csv',
 'germany_los_counterfactual_scenarios_2024.csv',
 'germany_los_gap_change_summary.csv']

In [115]:
los_scenarios = pd.read_csv(
    processed_dir
    / "germany_los_counterfactual_scenarios_2024.csv"
)

print(los_scenarios.shape)
print(los_scenarios.columns.tolist())

display(los_scenarios)

(6, 5)
['benchmark', 'benchmark_los', 'illustrative_nights', 'additional_nights_vs_actual', 'additional_nights_pct']


,benchmark,benchmark_los,illustrative_nights,additional_nights_vs_actual,additional_nights_pct
0,Germany Actual,2.265983,8.479264e+07,0.000000e+00,0.000000
1,EU27 Median,2.490083,9.317841e+07,8.385770e+06,9.889738
2,France,2.577033,9.643208e+07,1.163944e+07,13.726946
3,Netherlands,2.894230,1.083015e+08,2.350887e+07,27.725134
4,Austria,3.298724,1.234376e+08,3.864494e+07,45.575818
5,Weighted EU27,3.340360,1.249956e+08,4.020298e+07,47.413287


In [116]:
print(los_scenarios.dtypes)

display(
    los_scenarios
    .isna()
    .sum()
    .to_frame("missing_values")
)

benchmark                          str
benchmark_los                  float64
illustrative_nights            float64
additional_nights_vs_actual    float64
additional_nights_pct          float64
dtype: object


,missing_values
benchmark,0
benchmark_los,0
illustrative_nights,0
additional_nights_vs_actual,0
additional_nights_pct,0


In [117]:
los_scenarios_sql = los_scenarios.copy()

los_scenarios_sql["analysis_year"] = 2024
los_scenarios_sql["scenario_type"] = "Counterfactual Benchmark"

los_scenarios_sql["illustrative_nights"] = (
    los_scenarios_sql["illustrative_nights"]
    .round()
    .astype(int)
)

los_scenarios_sql["additional_nights_vs_actual"] = (
    los_scenarios_sql["additional_nights_vs_actual"]
    .round()
    .astype(int)
)

los_scenarios_sql = los_scenarios_sql[
    [
        "analysis_year",
        "benchmark",
        "benchmark_los",
        "illustrative_nights",
        "additional_nights_vs_actual",
        "additional_nights_pct",
        "scenario_type"
    ]
].copy()

print(los_scenarios_sql.shape)
print(los_scenarios_sql.dtypes)

display(los_scenarios_sql)

(6, 7)
analysis_year                    int64
benchmark                          str
benchmark_los                  float64
illustrative_nights              int64
additional_nights_vs_actual      int64
additional_nights_pct          float64
scenario_type                      str
dtype: object


,analysis_year,benchmark,benchmark_los,illustrative_nights,additional_nights_vs_actual,additional_nights_pct,scenario_type
0,2024,Germany Actual,2.265983,84792636,0,0.000000,Counterfactual Benchmark
1,2024,EU27 Median,2.490083,93178406,8385770,9.889738,Counterfactual Benchmark
2,2024,France,2.577033,96432075,11639439,13.726946,Counterfactual Benchmark
3,2024,Netherlands,2.894230,108301508,23508872,27.725134,Counterfactual Benchmark
4,2024,Austria,3.298724,123437574,38644938,45.575818,Counterfactual Benchmark
5,2024,Weighted EU27,3.340360,124995612,40202976,47.413287,Counterfactual Benchmark


In [118]:
print(
    "Duplicate year-benchmark rows:",
    los_scenarios_sql[
        ["analysis_year", "benchmark"]
    ]
    .duplicated()
    .sum()
)

print(
    "Missing values:",
    los_scenarios_sql.isna().sum().sum()
)

Duplicate year-benchmark rows: 0
Missing values: 0


In [119]:
display(
    los_scenarios_sql[
        los_scenarios_sql["benchmark"] == "Germany Actual"
    ]
)

,analysis_year,benchmark,benchmark_los,illustrative_nights,additional_nights_vs_actual,additional_nights_pct,scenario_type
0,2024,Germany Actual,2.265983,84792636,0,0.0,Counterfactual Benchmark


In [120]:
rows_inserted = los_scenarios_sql.to_sql(
    name="analysis_los_scenario",
    con=engine,
    if_exists="append",
    index=False
)

print("Rows inserted:", rows_inserted)

Rows inserted: 6


In [121]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                scenario_id,
                analysis_year,
                benchmark,
                benchmark_los,
                illustrative_nights,
                additional_nights_vs_actual,
                additional_nights_pct,
                scenario_type
            FROM analysis_los_scenario
            ORDER BY benchmark_los;
        """)
    )

    for row in result:
        print(row)

(1, 2024, 'Germany Actual', Decimal('2.2660'), 84792636, 0, Decimal('0.0000'), 'Counterfactual Benchmark')
(2, 2024, 'EU27 Median', Decimal('2.4901'), 93178406, 8385770, Decimal('9.8897'), 'Counterfactual Benchmark')
(3, 2024, 'France', Decimal('2.5770'), 96432075, 11639439, Decimal('13.7269'), 'Counterfactual Benchmark')
(4, 2024, 'Netherlands', Decimal('2.8942'), 108301508, 23508872, Decimal('27.7251'), 'Counterfactual Benchmark')
(5, 2024, 'Austria', Decimal('3.2987'), 123437574, 38644938, Decimal('45.5758'), 'Counterfactual Benchmark')
(6, 2024, 'Weighted EU27', Decimal('3.3404'), 124995612, 40202976, Decimal('47.4133'), 'Counterfactual Benchmark')


In [122]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                COUNT(*) AS scenarios,
                COUNT(DISTINCT benchmark) AS unique_benchmarks,
                MIN(analysis_year) AS min_year,
                MAX(analysis_year) AS max_year
            FROM analysis_los_scenario;
        """)
    )

    print(result.fetchone())

(6, 6, 2024, 2024)


In [123]:
outlook_files = sorted(
    [
        file.name
        for file in processed_dir.glob("*")
        if "2026" in file.name.lower()
        and (
            "projection" in file.name.lower()
            or "outlook" in file.name.lower()
            or "scenario" in file.name.lower()
            or "actual" in file.name.lower()
        )
    ]
)

outlook_files

['germany_2026_actual_projection_combined.csv',
 'germany_2026_baseline_scenario_summary.csv',
 'germany_2026_h2_seasonal_naive_projection.csv']

In [124]:
germany_outlook = pd.read_csv(
    processed_dir
    / "germany_2026_actual_projection_combined.csv"
)

print(germany_outlook.shape)
print(germany_outlook.columns.tolist())

display(germany_outlook)

(12, 5)
['date', 'year', 'month', 'foreign_nights_value', 'data_type']


,date,year,month,foreign_nights_value,data_type
0,2026-01-01,2026,1,4277203.0,Actual
1,2026-02-01,2026,2,5078278.0,Actual
2,2026-03-01,2026,3,5197078.0,Actual
3,2026-04-01,2026,4,6596278.0,Actual
4,2026-05-01,2026,5,7446648.0,Actual
5,2026-06-01,2026,6,7653746.0,Actual
6,2026-07-01,2026,7,10358796.0,Seasonal Naive Projection
7,2026-08-01,2026,8,9689151.0,Seasonal Naive Projection
8,2026-09-01,2026,9,7674317.0,Seasonal Naive Projection
9,2026-10-01,2026,10,7195093.0,Seasonal Naive Projection


In [125]:
print(germany_outlook.dtypes)

display(
    germany_outlook
    .isna()
    .sum()
    .to_frame("missing_values")
)

date                        str
year                      int64
month                     int64
foreign_nights_value    float64
data_type                   str
dtype: object


,missing_values
date,0
year,0
month,0
foreign_nights_value,0
data_type,0


In [126]:
print(germany_outlook.dtypes)

display(
    germany_outlook
    .isna()
    .sum()
    .to_frame("missing_values")
)

date                        str
year                      int64
month                     int64
foreign_nights_value    float64
data_type                   str
dtype: object


,missing_values
date,0
year,0
month,0
foreign_nights_value,0
data_type,0


In [127]:
print(
    germany_outlook["data_type"]
    .value_counts(dropna=False)
)

data_type
Actual                       6
Seasonal Naive Projection    6
Name: count, dtype: int64


In [128]:
germany_outlook_sql = germany_outlook.copy()

germany_outlook_sql["period_date"] = pd.to_datetime(
    germany_outlook_sql["date"]
)

germany_outlook_sql["foreign_nights_value"] = (
    germany_outlook_sql["foreign_nights_value"]
    .round()
    .astype(int)
)

germany_outlook_sql["observation_status"] = (
    germany_outlook_sql["data_type"]
    .map({
        "Actual": "Observed",
        "Seasonal Naive Projection": "Projected"
    })
)

germany_outlook_sql = germany_outlook_sql[
    [
        "period_date",
        "year",
        "month",
        "foreign_nights_value",
        "data_type",
        "observation_status"
    ]
].copy()

print(germany_outlook_sql.shape)
print(germany_outlook_sql.dtypes)

display(germany_outlook_sql)

(12, 6)
period_date             datetime64[us]
year                             int64
month                            int64
foreign_nights_value             int64
data_type                          str
observation_status                 str
dtype: object


,period_date,year,month,foreign_nights_value,data_type,observation_status
0,2026-01-01,2026,1,4277203,Actual,Observed
1,2026-02-01,2026,2,5078278,Actual,Observed
2,2026-03-01,2026,3,5197078,Actual,Observed
3,2026-04-01,2026,4,6596278,Actual,Observed
4,2026-05-01,2026,5,7446648,Actual,Observed
5,2026-06-01,2026,6,7653746,Actual,Observed
6,2026-07-01,2026,7,10358796,Seasonal Naive Projection,Projected
7,2026-08-01,2026,8,9689151,Seasonal Naive Projection,Projected
8,2026-09-01,2026,9,7674317,Seasonal Naive Projection,Projected
9,2026-10-01,2026,10,7195093,Seasonal Naive Projection,Projected


In [129]:
print(
    germany_outlook_sql[
        ["data_type", "observation_status"]
    ]
    .value_counts()
)

print(
    "Duplicate dates:",
    germany_outlook_sql["period_date"]
    .duplicated()
    .sum()
)

print(
    "Missing values:",
    germany_outlook_sql
    .isna()
    .sum()
    .sum()
)

data_type                  observation_status
Actual                     Observed              6
Seasonal Naive Projection  Projected             6
Name: count, dtype: int64
Duplicate dates: 0
Missing values: 0


In [130]:
display(
    germany_outlook_sql
    .groupby(
        ["data_type", "observation_status"],
        as_index=False
    )
    .agg(
        months=("month", "count"),
        foreign_nights=("foreign_nights_value", "sum")
    )
)

,data_type,observation_status,months,foreign_nights
0,Actual,Observed,6,36249231
1,Seasonal Naive Projection,Projected,6,47091061


In [131]:
print(
    "2026 full-year seasonal-naive scenario:",
    germany_outlook_sql["foreign_nights_value"].sum()
)

2026 full-year seasonal-naive scenario: 83340292


In [132]:
rows_inserted = germany_outlook_sql.to_sql(
    name="analysis_germany_tourism_outlook",
    con=engine,
    if_exists="append",
    index=False
)

print("Rows inserted:", rows_inserted)

Rows inserted: 12


In [133]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                observation_status,
                COUNT(*) AS months,
                SUM(foreign_nights_value) AS foreign_nights
            FROM analysis_germany_tourism_outlook
            GROUP BY observation_status
            ORDER BY observation_status;
        """)
    )

    for row in result:
        print(row)

('Observed', 6, Decimal('36249231'))
('Projected', 6, Decimal('47091061'))


In [134]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                COUNT(*) AS total_months,
                MIN(period_date) AS first_month,
                MAX(period_date) AS last_month,
                SUM(foreign_nights_value) AS full_year_scenario_nights,
                SUM(
                    CASE
                        WHEN month <= 6
                             AND observation_status <> 'Observed'
                        THEN 1
                        ELSE 0
                    END
                ) AS h1_status_errors,
                SUM(
                    CASE
                        WHEN month >= 7
                             AND observation_status <> 'Projected'
                        THEN 1
                        ELSE 0
                    END
                ) AS h2_status_errors
            FROM analysis_germany_tourism_outlook;
        """)
    )

    print(result.fetchone())

(12, datetime.date(2026, 1, 1), datetime.date(2026, 12, 1), Decimal('83340292'), Decimal('0'), Decimal('0'))


In [135]:
germany_outlook_sql = germany_outlook.copy()

germany_outlook_sql["period_date"] = pd.to_datetime(
    germany_outlook_sql["date"]
)

germany_outlook_sql["foreign_nights_value"] = (
    germany_outlook_sql["foreign_nights_value"]
    .round()
    .astype(int)
)

germany_outlook_sql["observation_status"] = (
    germany_outlook_sql["data_type"]
    .map({
        "Actual": "Observed",
        "Seasonal Naive Projection": "Projected"
    })
)

germany_outlook_sql = germany_outlook_sql[
    [
        "period_date",
        "year",
        "month",
        "foreign_nights_value",
        "data_type",
        "observation_status"
    ]
].copy()

print(germany_outlook_sql.shape)
print(germany_outlook_sql.dtypes)

display(germany_outlook_sql)

(12, 6)
period_date             datetime64[us]
year                             int64
month                            int64
foreign_nights_value             int64
data_type                          str
observation_status                 str
dtype: object


,period_date,year,month,foreign_nights_value,data_type,observation_status
0,2026-01-01,2026,1,4277203,Actual,Observed
1,2026-02-01,2026,2,5078278,Actual,Observed
2,2026-03-01,2026,3,5197078,Actual,Observed
3,2026-04-01,2026,4,6596278,Actual,Observed
4,2026-05-01,2026,5,7446648,Actual,Observed
5,2026-06-01,2026,6,7653746,Actual,Observed
6,2026-07-01,2026,7,10358796,Seasonal Naive Projection,Projected
7,2026-08-01,2026,8,9689151,Seasonal Naive Projection,Projected
8,2026-09-01,2026,9,7674317,Seasonal Naive Projection,Projected
9,2026-10-01,2026,10,7195093,Seasonal Naive Projection,Projected


In [136]:
print(
    germany_outlook_sql[
        ["data_type", "observation_status"]
    ]
    .value_counts()
)

print(
    "Duplicate dates:",
    germany_outlook_sql["period_date"]
    .duplicated()
    .sum()
)

print(
    "Missing values:",
    germany_outlook_sql
    .isna()
    .sum()
    .sum()
)

data_type                  observation_status
Actual                     Observed              6
Seasonal Naive Projection  Projected             6
Name: count, dtype: int64
Duplicate dates: 0
Missing values: 0


In [137]:
display(
    germany_outlook_sql
    .groupby(
        ["data_type", "observation_status"],
        as_index=False
    )
    .agg(
        months=("month", "count"),
        foreign_nights=("foreign_nights_value", "sum")
    )
)

,data_type,observation_status,months,foreign_nights
0,Actual,Observed,6,36249231
1,Seasonal Naive Projection,Projected,6,47091061


In [138]:
print(
    "2026 full-year seasonal-naive scenario:",
    germany_outlook_sql["foreign_nights_value"].sum()
)

2026 full-year seasonal-naive scenario: 83340292


In [140]:
rows_inserted = germany_outlook_sql.to_sql(
    name="analysis_germany_tourism_outlook",
    con=engine,
    if_exists="append",
    index=False
)

print("Rows inserted:", rows_inserted)

DatabaseError: Execution failed on sql 'INSERT INTO analysis_germany_tourism_outlook (period_date, year, month, foreign_nights_value, data_type, observation_status) VALUES (:period_date, :year, :month, :foreign_nights_value, :data_type, :observation_status)': (pymysql.err.IntegrityError) (1062, "Duplicate entry '2026-01-01' for key 'analysis_germany_tourism_outlook.PRIMARY'")
[SQL: INSERT INTO analysis_germany_tourism_outlook (period_date, year, month, foreign_nights_value, data_type, observation_status) VALUES (%(period_date)s, %(year)s, %(month)s, %(foreign_nights_value)s, %(data_type)s, %(observation_status)s)]
[parameters: [{'period_date': datetime.datetime(2026, 1, 1, 0, 0), 'year': 2026, 'month': 1, 'foreign_nights_value': 4277203, 'data_type': 'Actual', 'observation_status': 'Observed'}, {'period_date': datetime.datetime(2026, 2, 1, 0, 0), 'year': 2026, 'month': 2, 'foreign_nights_value': 5078278, 'data_type': 'Actual', 'observation_status': 'Observed'}, {'period_date': datetime.datetime(2026, 3, 1, 0, 0), 'year': 2026, 'month': 3, 'foreign_nights_value': 5197078, 'data_type': 'Actual', 'observation_status': 'Observed'}, {'period_date': datetime.datetime(2026, 4, 1, 0, 0), 'year': 2026, 'month': 4, 'foreign_nights_value': 6596278, 'data_type': 'Actual', 'observation_status': 'Observed'}, {'period_date': datetime.datetime(2026, 5, 1, 0, 0), 'year': 2026, 'month': 5, 'foreign_nights_value': 7446648, 'data_type': 'Actual', 'observation_status': 'Observed'}, {'period_date': datetime.datetime(2026, 6, 1, 0, 0), 'year': 2026, 'month': 6, 'foreign_nights_value': 7653746, 'data_type': 'Actual', 'observation_status': 'Observed'}, {'period_date': datetime.datetime(2026, 7, 1, 0, 0), 'year': 2026, 'month': 7, 'foreign_nights_value': 10358796, 'data_type': 'Seasonal Naive Projection', 'observation_status': 'Projected'}, {'period_date': datetime.datetime(2026, 8, 1, 0, 0), 'year': 2026, 'month': 8, 'foreign_nights_value': 9689151, 'data_type': 'Seasonal Naive Projection', 'observation_status': 'Projected'}  ... displaying 10 of 12 total bound parameter sets ...  {'period_date': datetime.datetime(2026, 11, 1, 0, 0), 'year': 2026, 'month': 11, 'foreign_nights_value': 5590732, 'data_type': 'Seasonal Naive Projection', 'observation_status': 'Projected'}, {'period_date': datetime.datetime(2026, 12, 1, 0, 0), 'year': 2026, 'month': 12, 'foreign_nights_value': 6582972, 'data_type': 'Seasonal Naive Projection', 'observation_status': 'Projected'}]]
(Background on this error at: https://sqlalche.me/e/20/gkpj)

In [141]:
with engine.connect() as connection:
    result = connection.execute(
        text("""
            SELECT
                COUNT(*) AS total_rows,
                COUNT(DISTINCT period_date) AS unique_dates,
                SUM(CASE
                    WHEN observation_status = 'Observed'
                    THEN 1 ELSE 0
                END) AS observed_months,
                SUM(CASE
                    WHEN observation_status = 'Projected'
                    THEN 1 ELSE 0
                END) AS projected_months,
                SUM(foreign_nights_value) AS total_nights
            FROM analysis_germany_tourism_outlook;
        """)
    )

    print(result.fetchone())

(12, 12, Decimal('6'), Decimal('6'), Decimal('83340292'))
